src/nvim/mark_extended.c

```diff
@@ -0,0 +1,305 @@
+{
+  return extmark_neighbour(buf, ns, pos, 0);
+}
+
+ExtmarkArray *extmark_nextrange(buf_T *buf, char *ns, pos_T *lower, pos_T *upper)
+{
+  return extmark_neighbour_range(buf, ns, lower, upper, 1);
+}
+
+/* Returns the position of marks between a range, */
+/* marks found at the start or end index will be included, */
+/* if upper.lnum or upper.col are negative the buffer */
+/* will be searched to the end, */
+/* go_forward can be set to control the order of the array */
+static ExtmarkArray *extmark_neighbour_range(buf_T *buf, char *ns, pos_T *lower, pos_T *upper, bool go_forward)
+{
+  ExtmarkArray *array = (ExtmarkArray *)xmalloc(sizeof(ExtmarkArray));
+  kv_init(*array);
+  int cmp;
+
+  FOR_EXTMARKS_IN_NS(buf, ns)
+    if (pos_cmp(*lower, *upper) == 1) {
+      break;
+    }
+    cmp = pos_cmp(*lower, extmark->mark);
+    if (cmp != 1) {
+      kv_push(*array, extmark);
+    }
+    else if (cmp == 1) {
+      break;
+    }
+  END_FOR_EXTMARKS_IN_NS
+
+  /* Swap the elements in the array */
+  /* if (!go_forward) { */
+      /* ExtendedMark holder; */
+      /* int n = (int)sizeof(array); */
+      /* int end = n-1; */
+      /* for (int i = 0; i < n/2; i++) { */
+          /* holder = kv_A(*array, i); */
+          /* array[i] = array[end]; */
+          /* array[end] = holder; */
+          /* end --; */
+      /* } */
+  /* } */
+  return array;
+}
+
+/* Returns the mark, nearest to the passed in mark*/
+static ExtendedMark *extmark_neighbour(buf_T *buf, char *ns, pos_T *input, bool go_forward)
+{
+  int cmp;
+  if (go_forward) {
+    FOR_ALL_EXTMARKS(buf)
+      cmp = pos_cmp(extmark->mark, *input);
+      if (cmp == 1) {
+        return extmark;
+      }
+    END_FOR_ALL_EXTMARKS
+  } else {
+    FOR_ALL_EXTMARKS_WITH_PREV(buf)
+      cmp = pos_cmp(extmark->mark, *input);
+      if (cmp == 1) {
+        return prev;
+      }
+    END_FOR_ALL_EXTMARKS_WITH_PREV
+  }
+  return NULL;
+}
+static bool extmark_create(buf_T *buf, char* ns, char *name, linenr_T row, colnr_T col)
+{
+  /* Initialize as first mark on this buffer*/
+  if (!buf->b_extmarks) {
+    buf->b_extmarks = kb_init(extmarks, KB_DEFAULT_SIZE);
+    buf->b_extmark_ns = pmap_new(cstr_t)();
+  }
+  ExtmarkNamespace *ns_obj = pmap_get(cstr_t)(buf->b_extmark_ns, (cstr_t)ns);
+  /* Initialize a new namespace */
+  /* if (!ns_map || !kh_size(ns_map->table)) { */
+  ns = xstrdup(ns);
+  if (!ns_obj) {
+    ns_obj = (ExtmarkNamespace *)xmalloc(sizeof(ExtmarkNamespace));
+    ns_obj->map = pmap_new(cstr_t)();
+    ns_obj->tree = kb_init(extmarks, KB_DEFAULT_SIZE);
+    pmap_put(cstr_t)(buf->b_extmark_ns, (cstr_t)ns, ns_obj);
+  }
+
+  name = xstrdup(name);
+  ExtendedMark *extmark = (ExtendedMark *)xmalloc(sizeof(ExtendedMark));
+  /* extmark->mark = gen_relative(buf, ns, row, col); */
+  extmark->mark.lnum = row;
+  extmark->mark.col = col;
+  extmark->name = name;
+
+  kb_put(extmarks, buf->b_extmarks, extmark);
+  kb_put(extmarks, ns_obj->tree, extmark);
```

Comment:
not sure what the second put is for. Seems redundant to a have a tree per namespace.